src/dbos-executor.ts

```diff
@@ -831,6 +886,135 @@ export class DBOSExecutor implements DBOSExecutorContext {
     }, ...args)).getResult();
   }
 
+  async callTransactionFunction<T extends unknown[], R>(
+    txn: Transaction<T, R>, clsinst: ConfiguredInstance | null, wfCtx: WorkflowContextImpl, ...args: T
+  ): Promise<R> {
+    const txnInfo = this.getTransactionInfo(txn as Transaction<unknown[], unknown>);
+    if (txnInfo === undefined) {
+      throw new DBOSNotRegisteredError(txn.name);
+    }
+    const readOnly = txnInfo.config.readOnly ?? false;
+    let retryWaitMillis = 1;
+    const backoffFactor = 1.5;
+    const maxRetryWaitMs = 2000; // Maximum wait 2 seconds.
+    const funcId = wfCtx.functionIDGetIncrement();
+    const span: Span = this.tracer.startSpan(
+      txn.name,
+      {
+        operationUUID: wfCtx.workflowUUID,
+        operationType: OperationType.TRANSACTION,
+        authenticatedUser: wfCtx.authenticatedUser,
+        assumedRole: wfCtx.assumedRole,
+        authenticatedRoles: wfCtx.authenticatedRoles,
+        readOnly: readOnly,
+        isolationLevel: txnInfo.config.isolationLevel,
+      },
+      wfCtx.span,
+    );
+
+    while (true) {
+      let txn_snapshot = "invalid";
+      const workflowUUID = wfCtx.workflowUUID;
+      const wrappedTransaction = async (client: UserDatabaseClient): Promise<R> => {
+        const tCtxt = new TransactionContextImpl(
+          this.userDatabase.getName(), client, wfCtx,
+          span, this.logger, funcId, txn.name);
+
+        // If the UUID is preset, it is possible this execution previously happened. Check, and return its original result if it did.
+        // Note: It is possible to retrieve a generated ID from a workflow handle, run a concurrent execution, and cause trouble for yourself. We recommend against this.
+        if (wfCtx.presetUUID) {
+          const check: BufferedResult = await wfCtx.checkTxExecution<R>(client, funcId);
+          txn_snapshot = check.txn_snapshot;
+          if (check.output !== dbosNull) {
+            tCtxt.span.setAttribute("cached", true);
+            tCtxt.span.setStatus({ code: SpanStatusCode.OK });
+            this.tracer.endSpan(tCtxt.span);
+            return check.output as R;
+          }
+        } else {
+          // Collect snapshot information for read-only transactions and non-preset UUID transactions, if not already collected above
+          txn_snapshot = await wfCtx.retrieveTxSnapshot(client);
+        }
+
+        // For non-read-only transactions, flush the result buffer.
+        if (!readOnly) {
+          await wfCtx.flushResultBuffer(client);
+        }
+
+        // Execute the user's transaction.
+        let cresult: R | undefined;
+        if (txnInfo.registration.passContext) {
+          await runWithTransactionContext(tCtxt, async ()=> {
+            cresult = await txn.call(clsinst, tCtxt, ...args);
+          });
+        }
+        else {
+          await runWithTransactionContext(tCtxt, async ()=> {
+            const tf = txn as unknown as (...args: T)=>Promise<R>;
```

Comment:
Should we define a `ContextFreeTransaction` type? :) Or overkill?

Reply:
I do not know why you would do that